src/transition.ts

```diff
@@ -0,0 +1,54 @@
+import * as t from 'tcomb';
+import shallowEqual from './shallowEqual';
+import { BehaviorSubject } from 'rxjs/BehaviorSubject';
+import * as debug from 'debug';
+
+type StateT = { view: string };
+export type TransitionFunctionFunction<S extends StateT> = (s: S) => S;
+export type TransitionFunctionPatch<S extends StateT> = { [k in keyof S]?: S[k] | null };
+export type TransitionFunction<S extends StateT> = (
+  transition: TransitionFunctionFunction<S> | TransitionFunctionPatch<S>
+) => void;
+
+const log = debug('state:run');
+
+type MakeTransitionParams<S extends StateT> = {
+  stateSubject: BehaviorSubject<S>;
+  transitionReducer: TransitionFunctionFunction<S>;
+  stateType: t.Interface<S>;
+  syncToBrowser: (oldState: S, newState: S) => void;
+};
+export default function makeTransition<S extends StateT>({
+  stateSubject,
+  transitionReducer,
+  stateType,
+  syncToBrowser
+}: MakeTransitionParams<S>): TransitionFunction<S> {
+  return _transition => {
+    log('transition:', _transition);
+
+    const isTransitionFunction = t.Function.is(_transition);
+    const transitionFn = isTransitionFunction
+      ? _transition as TransitionFunctionFunction<S> // TODO(typo): cast
+      : ((() => _transition) as any) as TransitionFunctionFunction<S>; // TODO(typo): cast + removed `t.Object(` check
+
+    const state = stateSubject.value;
+    const patch = transitionFn(state);
+
+    log('state is:', state);
+    log('patch is:', patch);
+    const shouldReplace = isTransitionFunction;
+    const newState = stateType(
+      // TODO(typo): double check, it was:
+      // transitionReducer(shouldReplace ? { ...patch } : { ...state, ...patch })
+      transitionReducer(shouldReplace ? Object.assign({}, patch) : Object.assign({}, state, patch))
```

Comment:
Not with spread types in general, but `[ts] Spread types may only be created from object types`.
I think this is just because the type parameter `S` doesn't extend `{}` directly here.

Added a comment to clarify